lib/edamam.rb

```diff
@@ -0,0 +1,191 @@
+module Edamam
+  API_HOST = 'api.edamam.com'.freeze
+  APP_ID = ENV.fetch('EDAMAM_APP_ID')
+  APP_KEY = ENV.fetch('EDAMAM_APP_KEY')
+  MEASUREMENT_URI_PREFIX = 'http://www.edamam.com/ontologies/edamam.owl#Measure_'.freeze
+  NUTRITION_DATA_PATH = '/api/food-database/nutrients'.freeze
+  UPC_LOOKUP_PATH = '/api/food-database/parser'.freeze
+
+  class << self
+    def get_data_from_upc(upc, nutrients: false)
+      upc_lookup_data = send_request(
+        host: API_HOST,
+        method: :get,
+        name: 'UPC lookup',
+        path: UPC_LOOKUP_PATH,
+        query: { upc: upc }
+      )&.dig('hints', 0)
+
+      return unless upc_lookup_data
+
+      food_data = upc_lookup_data['food']
+
+      measurement_data = upc_lookup_data['measures']
+      measurement_units = measurement_data.map do |measure|
+        [measure['label'], measure['uri'].match(/^#{MEASUREMENT_URI_PREFIX}([a-z_]+)$/)[1]]
+      end.compact.to_h
+
+      edamam_id = food_data['foodId']
+      volumetric = measurement_units.key?('Tablespoon')
+      response = send_request(
+        body: {
+          ingredients: [
+            {
+              quantity: 1,
+              measureURI: measurement_uri(volumetric ? 'tablespoon' : 'gram'),
+              foodId: edamam_id
+            }
+          ]
+        }.to_json,
+        headers: { 'Content-Type': 'application/json' },
+        host: API_HOST,
+        method: :post,
+        name: 'UPC data',
+        path: NUTRITION_DATA_PATH
+      )
+      return unless response
+
+      if volumetric
+        grams_per_tablespoon = get_weight_from_nutrition_data_response(response)
+        return unless grams_per_tablespoon
+      end
+
+      if nutrients
+        response_nutrients = response['totalNutrients']
+
+        divisor = grams_per_tablespoon || 1
+        calories = cast_decimal(response_nutrients.dig(nutrient_mappings[:calories], 'quantity'))
+        calories /= divisor if calories
```

Comment:
Are there cases where we don't get calories back?

Reply:
Yup. There's another place in the response that seems to always have it but it is less precise and when the total nutrient value isn't there then that one will just incorrectly say 0 anyway.